source/adios2/toolkit/format/bp5/BP5Serializer.cpp

```diff
@@ -1333,19 +1333,27 @@ BP5Serializer::TimestepInfo BP5Serializer::CloseTimestep(int timestep,
     MBase->BitField = tmp;
     NewAttribute = false;
 
-    struct TimestepInfo Ret
-    {
-        Formats, Metadata, AttrData, CurDataBuffer
-    };
+    struct TimestepInfo Ret;
+    Ret.NewMetaMetaBlocks = Formats;
```

Comment:
@pnorbert I figured the issue. IBM XL does not do the RVO optimization meaning that returning the struct will call the destructoror of TimeStepInfo. This resulted in this the DataBuffer being freed before being used.

Reply:
Unfortunate.  That'll presumably require an internal API change to fix.  If nobody else does it, I'll look tomorrow.

Reply:
Find the commit that adds the fix to it in: https://github.com/ornladios/ADIOS2/pull/3610/commits/501278282a04712b490eccc7519e23f71f20a3a7

Reply:
That would be hard to backport since we change member attributes of a public class, we can do a backport of the fix by removing the delete DataBuffer in the destructor of TimestepInfo

Reply:
Not a public class, at least not in the sense that it is exposed to users.  If it is changed, only an ADIOS-internal ABI changes, not the user-facing ABI.

Reply:
@eisenhauer we cannot be sure, this is a tricky one, it is a ABI breaking change for users that potentially interact with more internal classes of ADIOS2 (Which probably doesnt exist). The problem is here: https://github.com/ornladios/ADIOS2/blob/8ce4ff9bd43091f76de5f12dac1a19d7ddd49be1/source/adios2/toolkit/format/bp5/BP5Serializer.h#LL110C1-L110C17

This method is publicly available meaning that an user that an existing built app that uses this method when using the new 2.9.1 ADIOS2 libraries ld.so would not be able runlink due to changes in the memory size of the return type of that method.

Probably the best course of action is to adapt this to be ABI compatible when backporting to release_29. Not that it would realistically break users adios2 installations but for avoiding problems when adding automated API/ABI checks is best to be strict with it.

How about having this merged to master and later backport this to the release_29 in its own PR.